### PR TITLE
H371: ISAB single-layer probe — REPLACE only idx 2 of 5-layer stack

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

**H371 — ISAB single-layer probe (REPLACE only layer idx 2, the middle of the 5-layer stack).** Bounded-blast-radius successor to H370.

H370 (PR #1564 CLOSED) tested ISAB on 3 of 5 layers (idx 1/2/3) and regressed +97bp past the close threshold (val 7.019%). Edward's diagnosis identified two confounds:

1. **Mixed-init pathology**: cold ISAB at idx 1/2/3 sandwiched between pretrained Transolver at idx 0 and idx 4 produces near-noise outputs the downstream pretrained layers cannot interpret. 3 epochs is far too few to re-learn 60% of the slice-mixing operator.
2. **Compression magnitude**: S=128 → M=32 is 4:1 (the PR-body 3:1 was based on S=96 from the spec, but actual config is S=128). Combined with 3-layer breadth, the total perturbation magnitude blew the basin.

**H371 isolates the ISAB operator at a single layer (idx 2)** to test whether it has *any* merit under the H336 warm-start recipe:
- 1 layer of 5 replaced → 4 of 5 layers preserved (only 20% of slice-mixing perturbed)
- M=32 inducing points per head (same as H370 — keep one variable at a time)
- Same warm-start recipe (Lion, 3-epoch cosine tail, lr 9e-5)

**The point of H371 is to bound the ISAB operator family in two informative ways:**
- **Failure mode**: If 1-layer ISAB at idx 2 ALSO blows the 6.05% threshold, the ISAB operator is genuinely incompatible with the H336 basin under the warm-start-and-fine-tune framework. We close the entire ISAB axis (no M=64, no M=8 follow-ups).
- **Partial-success mode**: If 1-layer ISAB lands at 5.9–6.0%, we have signal that scaling DOWN ISAB (rather than UP) is the right recipe. We open the M=64 less-compression probe and the from-scratch-ISAB option next.
- **Best-case mode**: If 1-layer ISAB lands at <5.8962% AND test_cal beats 5.7357%, we have a new SOTA from a single architectural swap. Unlikely but the test cost is trivial.

**Why edward and not another student**: edward has now run 4-for-4 loss-tier nulls (H338/H361/H364/H368) plus 1 architectural null (H370). The pre-committed close rule has saved ~10h of GPU on these attempts. Edward's diagnostic comment on H370 shows he understands the basin/recipe interaction better than anyone — he's the right operator for the H371 follow-up.

### 1. Re-use H370 ISAB implementation, change one flag

Your `ISABTransolverAttention` class from H370 stays as-is. Same per-head `[H=4, M=32, dim_head=128]` shape, same `trunc_normal_(std=0.02)` init, same MAB1 → MAB2 forward pass, same FFN structure.

**The only architectural change is the layer-selection flag:**

```bash
--use-isab-layers 2          # NOT "1,2,3" — single layer at idx 2 only
--isab-inducing-points 32    # same M=32 as H370 — controlling one variable
```

Verify on rank 0 first forward: print which layers are ISAB:
```
[ISAB] layer idx 0: TransolverAttention
[ISAB] layer idx 1: TransolverAttention
[ISAB] layer idx 2: ISABTransolverAttention  ← the only ISAB layer
[ISAB] layer idx 3: TransolverAttention
[ISAB] layer idx 4: TransolverAttention
```

### 2. Same warm-start recipe — Phase 1 3-epoch cosine tail

```bash
cd target/
torchrun --standalone --nproc_per_node=8 train.py \
  --epochs 16 --epochs-already-done 13 \
  --resume-from-wandb yw2a5dyl --resume-alias epoch-13 \
  --lr 9e-5 --lr-cosine-t-max 3 --lr-min 1e-6 \
  --optimizer lion --weight-decay 0.0005 --grad-clip-norm 0.5 \
  --model-layers 5 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --model-mlp-ratio 4 --use-qk-norm --use-surf-to-vol-xattn \
  --rff-num-features 16 --rff-init-sigmas 0.25,0.5,1.0,2.0,4.0 \
  --pos-encoding-mode string_separable --drop-path-max 0.1 \
  --tau-y-loss-weight 1.3 --tau-z-loss-weight 1.67 --mirror-augmentation \
  --batch-size 4 --train-surface-points 65536 --train-volume-points 65536 \
  --eval-surface-points 65536 --eval-volume-points 65536 \
  --save-every-epoch \
  --use-isab-layers 2 --isab-inducing-points 32 \
  --wandb-group h371-isab-1layer-idx2 \
  --wandb-name edward/h371-isab-m32-1layer-idx2-phase1
```

Phase 1 wall-clock: ~40 min for EP14 (same as H370 — M=32 < S=128 keeps the attention cost slightly lower than baseline). Total Phase 1 (EP14/EP15/EP16) ~2h on 8 GPUs.

### 3. Pre-committed close rule (STRICT — same as H370)

- If **EP14 val_primary > 6.05%**, close immediately. Do NOT run EP15/EP16. Post EP14 metrics and the close-rule trigger in your SENPAI-RESULT comment.
- If EP14 val_primary ≤ 6.05%, continue to EP15 and EP16, then run the standard H336/H342 TTA evaluation chain (K=5 + Student-t ν=4 + 8-res + mirror, calibrated) on the best EP.
- If best EP val_cal < 5.8962% AND test_cal < 5.7357%, this is a new SOTA. Otherwise the operator family is bounded but informative.

### 4. Smoke test (1 step on 1 GPU, before Phase 1 launch)

Confirm the single-layer ISAB replacement is wired correctly:
```bash
torchrun --standalone --nproc_per_node=1 train.py \
  --epochs 14 --epochs-already-done 13 \
  --resume-from-wandb yw2a5dyl --resume-alias epoch-13 \
  --use-isab-layers 2 --isab-inducing-points 32 \
  --max-train-steps 2 \
  --batch-size 2 \
  [...other args...]
```

Verify:
- ISABTransolverAttention is at idx 2 only
- `inducing_points` shape is `(4, 32, 128)` (one tensor, not 3)
- Backward step computes gradient on the inducing points (norm > 0)

### 5. SENPAI-RESULT marker (terminal or close-rule)

```
SENPAI-RESULT: {"terminal":true,"status":"<complete|closed>","pending_arms":false,"wandb_run_ids":["<run-id>"],"primary_metric":{"name":"val_abupt_calibrated","value":<number>},"test_metric":{"name":"test_abupt_calibrated","value":<number>}}
```

## Baseline

**Current SOTA (H342 PR #1526):**
- val_abupt_calibrated = **5.8962%**
- test_abupt_calibrated = **5.7357%**
- test_VP = 3.3751% | test_SP = 3.6124% | test_WSS = 6.6351% | test_WSS_z = 8.6122%
- Source: 3-cp output-space avg (ep13+ep14+ep15) of `yw2a5dyl`/`0gjfv45i` with H336 K=5+Student-t ν=4 × 8-res × mirror TTA + per-channel affine calibration
- W&B: `3icmxaqe`, `qgw0ix77`, `ijadzof0`

**Merge gate (AND-logic, strict):**
- val_abupt_calibrated < **5.8962%** AND test_abupt_calibrated < **5.7357%**

**Reproduce baseline checkpoint:** `yw2a5dyl:v0` (H336 EP13, Lion, canonical fine-tune root).

**Pre-committed close rule for H371:** EP14 val_primary > 6.05% → close (same threshold as H370/H368).

## Closed axes context

After H370 closure, we now have **19 closed axes** including 5 closed in the last edward streak (H338, H361, H364, H368, H370). Edward's diagnostic comment on H370 noted that 5 consecutive nulls share the recipe `warm-start yw2a5dyl EP13 + Lion + 3-epoch cosine` — the recipe itself may be the constraint, not the individual hypotheses. **H371 is the cheapest probe to test whether the ISAB operator family is fundamentally incompatible (1-layer also fails) or whether the H370 failure was about magnitude/depth (1-layer survives).** The answer to this binary question routes the entire next round of architectural attacks.

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT edward: H371 Phase 1 launched 2026-06-02 11:37Z.

**Implementation:**
- Cherry-picked the H370 ISAB commit (`ISABTransolverAttention`) onto the H371 branch — `model.py`/`train.py` are byte-identical to H370 except the layer-index flag.
- Architectural change: `--use-isab-layers 2` (single layer at idx 2). Layers 0/1/3/4 remain standard `TransolverAttention`.

**Smoke test (passed, 1-GPU, B=1, N=256 random):**
- Block placement (asserted): idx 0=Transolver, idx 1=Transolver, **idx 2=ISABTransolverAttention**, idx 3=Transolver, idx 4=Transolver.
- `inducing_points.shape = (4, 32, 128)` — **one tensor** at path `backbone.blocks.2.attention.inducing_points`.
- Backward step on the inducing points: grad sum-abs = 0.2968 (non-zero, gradient flows).
- Param delta: **+34,560 / 17.41M = +0.198%** (exactly 1/3 of H370's +0.595%, matching one of three ISAB layers).

**Phase 1 command (W&B group `h371-isab-1layer-idx2`, rank-0 run `mdzazf0k`):**
```
torchrun --standalone --nproc_per_node=8 train.py \
  --epochs 16 --epochs-already-done 13 \
  --resume-from-wandb yw2a5dyl --resume-alias epoch-13 \
  --lr 9e-5 --lr-cosine-t-max 3 --lr-min 1e-6 \
  --optimizer lion --weight-decay 0.0005 --grad-clip-norm 0.5 \
  --model-layers 5 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --model-mlp-ratio 4 --use-qk-norm --use-surf-to-vol-xattn \
  --rff-num-features 16 --rff-init-sigmas 0.25,0.5,1.0,2.0,4.0 \
  --pos-encoding-mode string_separable --drop-path-max 0.1 \
  --tau-y-loss-weight 1.3 --tau-z-loss-weight 1.67 --mirror-augmentation \
  --batch-size 4 --train-surface-points 65536 --train-volume-points 65536 \
  --eval-surface-points 65536 --eval-volume-points 65536 \
  --save-every-epoch \
  --use-isab-layers 2 --isab-inducing-points 32 \
  --wandb-group h371-isab-1layer-idx2 \
  --wandb-name edward/h371-isab-m32-1layer-idx2-phase1
```

**Deviation from H370:** layer flag is `2` (not `1,2,3`). All other flags match H370 exactly (Lion, lr 9e-5, cosine 3-epoch tail, τ_z=1.67, mirror, drop-path 0.1, M=32, S=128).

**Pre-committed close rule active:** EP14 val_primary > 6.05% → close immediately, no EP15/EP16. Will report EP14 metrics either way once validation lands (ETA ~50min). Step ~3 reached in <20s; rank-0 throughput nominal.

W&B rank-0: https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/mdzazf0k

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #1566 because ADVISOR CLOSE 2026-06-02 12:17Z — Finding V: `isab-single-layer-probe-null` (22nd closed axis). ISAB OPERATOR FAMILY CLOSED at the single-layer bound.

**Terminal data, W&B `mdzazf0k` (edward/h371-isab-m32-1layer-idx2-phase1-rank0, state=running, EP14 val just landed):**

| Channel | H371 EP14 | H336 EP13 source | Δ vs source | Δ vs H342 gate |
|---|---:|---:|---:|---:|
| **val_abupt** | **6.7149%** | 6.017% | **+0.70pp** | +1.18pp ❌ |
| val_SP | 4.4177% | ~3.78% | +0.64pp | — |
| val_VP | 4.5283% | ~3.45% | **+1.08pp** | — |
| val_WSS | 7.4403% | ~6.78% | +0.66pp | — |
| val_WSS_x | 6.5057% | — | sizable | — |
| val_WSS_y | 8.2410% | — | sizable | — |
| **val_WSS_z** | **9.8822%** | ~9.06% | **+0.82pp** | — (target channel WORSE) |

**Close-rule trigger:**
- Pre-committed: EP14 val > 6.05% → close
- Actual: 6.7149% → **+66bp above threshold** → triggers
- Step 2801, EP14 checkpoint path `outputs/drivaerml/run-mdzazf0k/checkpoint_ep14.pt` (confirmed EP14 marker, not mid-epoch)
- Train loss slope `train/slope/base_mse_loss/per_1k_steps = -0.018` → model is learning, but in a direction that doesn't recover H336 basin

**ISAB operator family conclusion (joint with H370):**

| Hypothesis | Depth | EP14 val | Δ vs source |
|---|---|---:|---:|
| H370 ISAB | 3-of-5 layers (idx 1,2,3) | 7.0188% | **+1.00pp** |
| H371 ISAB | 1-of-5 layers (idx 2 only) | 6.7149% | **+0.70pp** |

**Single-layer probe still gives +66bp catastrophic regression.** The ISAB operator (inducing-point compression with M=32 per-head) is fundamentally incompatible with the H336 EP13 fine-tune basin under 3-epoch cosine warm-start, **regardless of replacement depth**. Even one cold ISAB layer at idx 2 (the middle of the 5-layer Transolver stack) disrupts the calibrated slice-token interaction structure enough to lose +66bp on every channel including the +108bp VP untouched-channel collapse signature we saw on H370.

**Per-channel pattern**: All 5 channels regressed broadly (+64 to +108bp), with VP +108bp (the channel ISAB does not directly modify) showing the worst absolute regression. This is the same "basin lost all calibrated slice-token interaction structure" signature as H370 — confirming the mixed-init pathology is global, not local.

**Mechanism interpretation:** With M=32 inducing-point bottleneck (S=128 → M=32 → S=128 cross-attention), the ISAB layer cannot pass through the slice-token interaction structure that pretrained Transolver layers 0/1/3/4 expect on input. The 3-epoch cosine tail with Lion lr=9e-5 → 1e-6 provides insufficient gradient to re-learn either side of the cold ISAB layer to match.

**Implication for architectural-rewrite tier:** Inducing-point compression families (ISAB, MAB-based set transformers) are off the table for warm-start under EP13 basin. Future operator-replacement attempts must preserve the input-output token bijection — for example: sparse-local + global-pool hybrid (kNN-attention + one global token), or attention with low-rank value projection (no token-count change). Operator replacements that CHANGE the token count or compress through a bottleneck are now empirically excluded.

**22 closed axes**. Edward now idle. Combined with fern + frieren idle (post-H360 / H369 close), researcher-agent now generating hypotheses for **3 idle students** rather than 2. Updated dispatch context: prioritize output-space transforms, auxiliary-loss heads, test-time training, hard-example case-level mining, z-flip equivariance — and ANY operator replacement that preserves token-count bijection (no compression).

Edward: clean execution again. The single-layer probe was the right next test — bounded the operator family in one Phase 1 instead of needing a depth sweep. New hypothesis assignment incoming.

SENPAI-RESULT: {"terminal":true,"status":"closed_close_rule","pending_arms":false,"wandb_run_ids":["mdzazf0k"],"primary_metric":{"name":"val_abupt_raw_ep14","value":6.7149},"test_metric":{"name":"not_evaluated_close_rule","value":-1}}.
